File: lib/domain/models/board_point.dart
/// A single intersection on the 9x10 Xiangqi grid.
///
/// `row` runs 0-9: row 0 is Red's back rank (bottom of the board as rendered),
/// row 9 is Black's back rank (top). `col` runs 0-8, left to right.
class BoardPoint {
  final int row;
  final int col;

  const BoardPoint(this.row, this.col);

  BoardPoint translate(int deltaRow, int deltaCol) =>
      BoardPoint(row + deltaRow, col + deltaCol);

  @override
  bool operator ==(Object other) =>
      other is BoardPoint && other.row == row && other.col == col;

  @override
  int get hashCode => Object.hash(row, col);

  @override
  String toString() => '($row,$col)';
}

File: backend/src/xiangqi/boardPoint.ts
/** A single intersection on the 9x10 Xiangqi grid. `row` 0-9 (0 = Red's
 * back rank), `col` 0-8. Plain, comparable-by-value via `pointKey`/`pointsEqual`
 * since TS objects don't get structural `==` the way the Dart `BoardPoint`
 * does. */
export interface BoardPoint {
  readonly row: number;
  readonly col: number;
}

export function point(row: number, col: number): BoardPoint {
  return { row, col };
}

export function translate(p: BoardPoint, deltaRow: number, deltaCol: number): BoardPoint {
  return { row: p.row + deltaRow, col: p.col + deltaCol };
}

export function pointsEqual(a: BoardPoint, b: BoardPoint): boolean {
  return a.row === b.row && a.col === b.col;
}

/** Stable string key for using BoardPoints in Sets/Maps. */
export function pointKey(p: BoardPoint): string {
  return `${p.row},${p.col}`;
}
